code/espurna/mcp23s08.cpp

```diff
@@ -0,0 +1,238 @@
+ *
+ * @return void
+ */
+void MCP23S08SetDirection(uint8_t pinNumber, uint8_t mode)
+{
+    uint8_t registerData = MCP23S08ReadRegister(IODIR);
+
+    if (INPUT == mode)
+    {
+        registerData |= (1 << pinNumber);
+    }
+    else
+    {
+        registerData &= ~(1 << pinNumber);
+    }
+
+    MCP23S08WriteRegister(IODIR, registerData);
+}
+ 
+/**
+ * @brief Read an expander MCP23S08 a register.
+ *
+ * @param address A register address.
+ *
+ * @return The data from the register.
+ */
+uint8_t MCP23S08ReadRegister(uint8_t address)
+{
+    _expTxData[0] = READ_CMD;
+    _expTxData[1] = address;
+
+    digitalWrite(MCP23S08_CS, LOW);
+    SPI.transferBytes(_expTxData, _expRxData, 3);
+    digitalWrite(MCP23S08_CS, HIGH);
+
+    return _expRxData[2];
+} 
+
+/**
+ * @brief Write data in expander MCP23S08 register.
+ *
+ * @param address A register address.
+ * @param data A byte for write.
+ *
+ * @return void.
+ */
+void MCP23S08WriteRegister(uint8_t address, uint8_t data)
+{
+    _expTxData[0] = WRITE_CMD;
+    _expTxData[1] = address;
+    _expTxData[2] = data;
+
+    digitalWrite(MCP23S08_CS, LOW);
+    SPI.transferBytes(_expTxData, _expRxData, 3);
+    digitalWrite(MCP23S08_CS, HIGH);
+} 
+
+/**
+ * @brief Set expander MCP23S08 pin state.
+ *
+ * @param pinNumber The number of pin to be set.
+ * @param state The pin state, true - 1, false - 0.
+ *
+ * @return void
+ */
+void MCP23S08SetPin(uint8_t pinNumber, bool state)
+{
+    uint8_t registerData = MCP23S08ReadRegister(OLAT);
+
+    if (state)
+    {
+        registerData |= (1 << pinNumber);
+    }
+    else
+    {
+        registerData &= ~(1 << pinNumber);
+    }
+
+    MCP23S08WriteRegister(OLAT, registerData);
+}
+
+/**
+ * @brief Get MCP23S08 pin state.
+ *
+ * @param pinNumber The number of pin to be get.
+ *
+ * @return State true - 1, false - 0.
+ */
+bool MCP23S08GetPin(uint8_t pinNumber)
+{
+    uint8_t registerData = MCP23S08ReadRegister(GPIO);
+
+    return registerData & (1 << pinNumber);
+}
+ 
+/**
+ * @brief Set relay new state.
+ *
+ * @param relayNumber Number of relay from 0 to RELAY_COUNT - 1. 0 - Relay1, 1 - Relay2 ...
+ * @param state New state of relay, true - On, false = Off.
+ *
+ * @return void
+ */
+void MCP23S08SetRelayState(uint8_t relayNumber, bool state)
+{
+    // Check if relayNumber is out of range - return.
+    if (relayNumber > DUMMY_RELAY_COUNT - 1)
+    {
+        return;
+    }
+    
+    MCP23S08SetPin(RELAY_PINS[relayNumber], state);
+}
+
+/**
+ * @brief Get opto in state.
+ *
+ * @param optoInNumber OptoIn number from 0 to MCP23S08_OPTOIN_COUNT - 1
+ *
+ * @return bool true - opto in is On, false is Off. If number is out of range - return false.
+ */
+bool MCP23S08GetOptoInState(uint8_t optoInNumber)
+{
+
+    // Check if optoInNumber is out of range - return false.
+    if (optoInNumber > MCP23S08_OPTOIN_COUNT - 1)
+    {
+        return false;
+    }
+
+    return !MCP23S08GetPin(optoInNumber);
```

Comment:
btw here is the reason for button config, button_default_high can actually with this inversed

Reply:
I know,

I reused most of the code from the prodinowifi vendor with minimal change.

https://github.com/kmpelectronics/Arduino/tree/master/ProDinoWiFiEsp

Should we diverge so much from their code?

Reply:
We just want getpin & setpin, since we implement button and relay abstractions ourselves.
Note the https://github.com/dpeddi-iot/espurna/pull/1 where it uses getpin directly as digitalRead().